lib/IRGen/GenProto.cpp

```diff
@@ -947,23 +947,35 @@ namespace {
 
 /// Return true if the witness table requires runtime instantiation to
 /// handle resiliently-added requirements with default implementations.
-static bool isResilientConformance(const NormalProtocolConformance *conformance) {
+bool IRGenModule::isResilientConformance(
+    const NormalProtocolConformance *conformance) {
   // If the protocol is not resilient, the conformance is not resilient
   // either.
   if (!conformance->getProtocol()->isResilient())
     return false;
 
-  // If the protocol is in the same module as the conformance, we're
-  // not resilient.
-  if (conformance->getDeclContext()->getParentModule()
-      == conformance->getProtocol()->getParentModule())
+  auto *conformanceModule = conformance->getDeclContext()->getParentModule();
+
+  // If the protocol and the conformance are both in the current module,
+  // they're not resilient.
+  if (conformanceModule == getSwiftModule() &&
+      conformanceModule == conformance->getProtocol()->getParentModule())
+    return false;
+
+  // If the protocol and the conformance are in the same module and the
+  // conforming type is not generic, they're not resilient.
+  //
+  // This is an optimization -- a conformance of a non-generic type cannot
+  // resiliently become dependent.
+  if (!conformance->getDeclContext()->isGenericContext() &&
+      conformanceModule == conformance->getProtocol()->getParentModule())
```

Comment:
This is a new assumption, right?  What does it buy us?  Can we access the conformance more efficiently or something?

Reply:
The new condition is stronger than the old one. We went from "a conformance is resilient if defining module is different from the protocol's module" to "a conformance is resilient if defining module is different from the protocol's module OR the type is generic and not defined in the current module".

It would also be correct to say "a conformance is resilient if defining module is different from the protocol's module OR the type is not defined in the current module", however that regressed performance because what used to be simple references to symbols in the stdlib now had to go through an accessor function in the runtime.

Reply:
Right, it's that final point that I was curious about.  Directly accessing global witness table symbols basically assumes that these non-resilient conformances are always eagerly initialized on load.  But it sounds like we're committed to that, which (1) is probably okay and (2) we may not have a choice about anymore anyway.

Assuming that this doesn't affect anything else besides the access strategy for the conformance, I'm okay with this.